crates/core/tedge/src/cli/connect/command.rs

```diff
@@ -567,3 +573,136 @@ fn get_common_mosquitto_config_file_path(
+            url, configured_url
+        ),
+        Err(_) => println!("Failed to get the connected tenant URL from Cumulocity.\n"),
+    }
+}
+
+fn get_connected_c8y_url(port: u16) -> Result<String, ConnectError> {
+    const C8Y_TOPIC_BUILTIN_JWT_TOKEN_UPSTREAM: &str = "c8y/s/uat";
+    const C8Y_TOPIC_BUILTIN_JWT_TOKEN_DOWNSTREAM: &str = "c8y/s/dat";
+    const CLIENT_ID: &str = "get_jwt_token_c8y";
+
+    let mut options = MqttOptions::new(CLIENT_ID, DEFAULT_HOST, port);
+    options.set_keep_alive(RESPONSE_TIMEOUT);
+
+    let (mut client, mut connection) = rumqttc::Client::new(options, 10);
+    let mut acknowledged = false;
+
+    client.subscribe(C8Y_TOPIC_BUILTIN_JWT_TOKEN_DOWNSTREAM, AtLeastOnce)?;
+
+    for event in connection.iter() {
+        match event {
+            Ok(Event::Incoming(Packet::SubAck(_))) => {
+                // We are ready to get the response, hence send the request
+                client.publish(C8Y_TOPIC_BUILTIN_JWT_TOKEN_UPSTREAM, AtLeastOnce, false, "")?;
+            }
+            Ok(Event::Incoming(Packet::PubAck(_))) => {
+                // The request has been sent
+                acknowledged = true;
+            }
+            Ok(Event::Incoming(Packet::Publish(response))) => {
+                // We got a response
+                let token = String::from_utf8(response.payload.to_vec()).unwrap();
+                let connected_url = decode_jwt_token(token.as_str())?;
+                return Ok(connected_url);
+            }
+            Ok(Event::Outgoing(Outgoing::PingReq)) => {
+                // No messages have been received for a while
+                println!("Local MQTT publish has timed out.");
+                break;
+            }
+            Ok(Event::Incoming(Incoming::Disconnect)) => {
+                eprintln!("ERROR: Disconnected");
+                break;
+            }
+            Err(err) => {
+                eprintln!("ERROR: {:?}", err);
+                break;
+            }
+            _ => {}
+        }
+    }
+
+    if acknowledged {
+        // The request has been sent but without a response
+        println!("\nThe request has been sent, however, no response.");
+        Err(ConnectError::TimeoutElapsedError)
+    } else {
+        // The request has not even been sent
+        println!("\nMake sure mosquitto is running.");
+        Err(ConnectError::TimeoutElapsedError)
+    }
+}
+
+fn decode_jwt_token(token: &str) -> Result<String, ConnectError> {
+    // JWT token format: <header>.<payload>.<signature>. Thus, we want only <paylaod>.
+    let payload = token
+        .split_terminator('.')
+        .nth(1)
+        .ok_or(ConnectError::InvalidJWTToken {
+            token: token.to_string(),
+        })?;
+
+    let decoded = base64::decode(payload)?;
+    let json: serde_json::Value = serde_json::from_slice(decoded.as_slice())?;
+
+    let tenant_url = json["iss"].as_str().ok_or(ConnectError::InvalidJWTToken {
+        token: token.to_string(),
+    })?;
```

Comment:
I'm likely wrong here, but using `Value` does not guarantee that ["iss"] will exist and therefore may lead to a panic?

Reply:
Oh, square brackets can be used but this returns `Value::Null` in cases where get would have returned None.
I suppose that as Null is not a string you you get `None` from `as_str` which then results in an error.

Reply:
This `as_str()` is implemented in `serde_json::Value`, which returns `Option<&str>`. https://docs.serde.rs/serde_json/value/enum.Value.html#method.as_str
If it returns `None`, `ok_or()` converts the `None` to `Err`.